[ALLUXIO-2207] Use a range query to limit data fetched by Swift InputStream

madanadit opened the pull request:
This addresses JIRA https://alluxio.atlassian.net/browse/ALLUXIO-2207.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11214/
Test PASSed.

calvinjia commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -11,11 +11,17 @@
 import org.javaswift.joss.model.Account;
 import org.javaswift.joss.model.StoredObject;
 
+import org.slf4j.Logger;
```
This can be combined with the previous `.org` import block

calvinjia commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -92,33 +99,44 @@ public long skip(long n) throws IOException {
       return 0;
     }
+
+    LOG.debug("Swift InputStream {}: skipping {} bytes", this.hashCode(), n);
```
Do we want to keep these debug messages?

calvinjia commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -92,33 +99,44 @@ public long skip(long n) throws IOException {
     StoredObject storedObject = mAccount.getContainer(mContainerName).getObject(mObjectPath);
     DownloadInstructions downloadInstructions  = new DownloadInstructions();
-    downloadInstructions.setRange(new ExcludeStartRange(mPos));
+    final long endPos = mPos + Configuration.getBytes(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT);
```
It would be more robust to get the blocksize bytes from the file, but currently I think this is fine.

calvinjia replied:
Also, what happens if the user tries to read more than the provisioned range?

madanadit replied:
you're right that it will fail to read if user tries to read more than provisioned. but when does that happen?. will alluxio not try to read from UFS in chunks of blocksize. 

although it is still true that opening a block for read and closing it without reading entire block is still inefficient.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11229/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11230/
Test PASSed.

calvinjia commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -92,33 +98,44 @@ public long skip(long n) throws IOException {
     closeStream();
     mPos += n;
     openStream();
+    //LOG.debug("Swift InputStream {}: done skipping", this.hashCode());
```
I think we should just remove it if we are going to comment it out.

calvinjia commented:
@madanadit This looks good aside from the question about reading beyond the range.

madanadit commented:
@calvinjia Thanks for the review!. I have modified the behavior to openStream again in case a chunk boundary has been reached.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11243/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11245/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11246/
Test PASSed.

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -102,23 +111,47 @@ public long skip(long n) throws IOException {
+  }
+
+  /**
+   * An object is read in chunks of clock size. Check if a boundary has been reached.
```
s/clock/block

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -102,23 +111,47 @@ public long skip(long n) throws IOException {
+
+  /**
+   * Block size for reading an object in chunks.
+   * @return block size in bytes
```
add empty line before `@return`

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -102,23 +111,47 @@ public long skip(long n) throws IOException {
+   */
+  private void checkStream() {
+    if (mPos % getBlockSize() == 0) {
+      mStream = null;
```
should we close the stream before resetting it?

madanadit replied:
that's correct. thanks.

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -63,6 +70,7 @@ public int read() throws IOException {
     int value = mStream.read();
     if (value != -1) { // valid data read
       mPos++;
+      checkStream();
```
I would replace this with:

```
if (mPos % getBlockSize() == 0) {
  closeStream()
}
```

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -83,6 +91,7 @@ public int read(byte[] b, int offset, int length) throws IOException {
     int read = mStream.read(b, offset, length);
     if (read != -1) {
       mPos += read;
+      checkStream();
```
ditto

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -102,23 +111,47 @@ public long skip(long n) throws IOException {
+  /**
+   * An object is read in chunks of clock size. Check if a boundary has been reached.
+   */
+  private void checkStream() {
```
the method name and comment are misleading, because the implementation not only checks the stream, but also resets it if the boundary was reached ... I would get rid of it and inline it where needed

jsimsa commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -102,23 +111,47 @@ public long skip(long n) throws IOException {
    * Closes the current stream.
    */
   private void closeStream() throws IOException {
+    LOG.debug("Swift InputStream {}: closing stream with pos {}", this.hashCode(), mPos);
```
s/with/at

jsimsa commented:
@madanadit did a pass and left some comments

madanadit commented:
@jsimsa The JOSS javadoc is out of date. AbstractRange has 'long' fields: https://github.com/javaswift/joss/blob/bf0b8e411a1022e167e034ba4c3482670d46b6fe/src/main/java/org/javaswift/joss/headers/object/range/AbstractRange.java. 

Unfortunately MidPartRange (https://github.com/javaswift/joss/blob/bf0b8e411a1022e167e034ba4c3482670d46b6fe/src/main/java/org/javaswift/joss/headers/object/range/MidPartRange.java) casts 'int' into 'long' fields. That is why I created this class.

I have refactored to store 'length' instead of 'endPos' in AbstractRange.length field, and modified the getTo method in SwiftRange accordingly.

madanadit commented:
@jsimsa @calvinjia I've addresses all the comments. Please have another look. Thanks!.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11262/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11263/
Test PASSed.

calvinjia commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -102,23 +115,39 @@ public long skip(long n) throws IOException {
     }
     mStream.close();
     mStream = null;
+    LOG.debug("Swift InputStream {}: stream closed", this.hashCode());
```
Let's remove the debugging statements? Or maybe just keep a few you think are especially valuable?

madanadit replied:
@calvinjia done! :-)

jsimsa commented:
LGTM

calvinjia commented on underfs/swift/src/main/java/alluxio/underfs/swift/SwiftInputStream.java:
```diff
@@ -83,6 +93,9 @@ public int read(byte[] b, int offset, int length) throws IOException {
       mPos += read;
+      if (mPos % getBlockSize() == 0) {
+        closeStream();
+      }
```
Will this work in the following scenario?
```
mPos = 10
blockSize = 20
Read 20 bytes
```

jsimsa replied:
scenario 1) wouldn't that result in a partial read of 10 bytes

scenario 2) my understanding is that the code maintains the following invariant: `mPos % blockSize == 0 => mStream == null`

calvinjia replied:
Yeah for scenario 1, but I think that is okay because we will return `10` since we didn't break any contracts. The caller would need to call read in a loop to get the full data.

You are right for scenario 2, but then when will we ever return `-1`? Will `openStream` do the right thing when the range is beyond the file size? Or will we get an exception from the Swift library?

madanadit replied:
@calvinjia 

scenario 1) read will do a partial read of 10 bytes and closeStream (sets mStream = null). next read operation will call openStream.

my understanding is that inputstream should always be called in a loop as the interface specifies that "An attempt is made to read as many as len bytes, but a smaller number may be read. The number of bytes actually read is returned as an integer."

scenario 2) 
case 1. a previous read() incremented the position to 20 => mStream will be set to 'null'; next read will call openStream. 
case 2. a skip() incremented the position to 20 => skip called openStream for range (20,40). read() does not return -1.

madanadit replied:
@calvinjia We will return -1 when storedObject.downloadObjectAsInputStream returns an empty stream for a range beyond file size. Let me verify this scenario. Thanks for the catch!.

calvinjia replied:
Oh also, what will the behavior be if the file is not perfectly block aligned. For example:
```
blockSize = 10
fileLength = 8
```
Will the Swift library give back a 8 length stream or do we need to something like `min(fileLength, endPos)` before putting it in the range?

jsimsa replied:
I would expect the Swift library to return 8, but it would be good to verify this.

@madanadit it would also be good to record the above mentioned invariant in the class documentation

madanadit replied:
verified: the swift library truncates the stream to fileLength (8 in this case) => no need for explicit min(fileLength, endPos).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/11264/
Test PASSed.

calvinjia commented:
LGTM

madanadit commented:
@calvinjia @jsimsa I mentioned the invariant as a comment. This should be good to go now. Thanks for the discussion.
